File: lib/basecamp_mcp/help_parser.rb
require "strscan"

module BasecampMcp
  # Parses `basecamp <group> <action> --help` output into a structured description
  # of positional arguments and flags. Used only by the generator — not on the
  # runtime hot path.
  module HelpParser
    module_function

    # Returns:
    #   {
    #     summary:    "Create a new todo in a project.",
    #     positional: [{name: "content", required: true, description: "Content"}, ...],
    #     flags:      [{name: "due", short: "d", type: "string", description: "..."}, ...],
    #   }
    def parse(help_text)
      {
        summary:    extract_summary(help_text),
        positional: extract_positional(help_text),
        flags:      extract_flags(help_text),
      }
    end

    def extract_summary(text)
      # First non-blank line(s) until USAGE.
      lines = []
      text.each_line do |line|
        break if line.strip == "USAGE"
        lines << line.strip unless line.strip.empty?
      end
      lines.join(" ")
    end

    def extract_positional(text)
      section = extract_section(text, "ARGUMENTS")
      return [] unless section

      section.lines.filter_map do |line|
        # Required: "  <name>  Description"
        # Optional: "  [name]  Description"
        if (m = line.match(/\A\s+<(\w[\w-]*)>\s+(.*)/))
          { "name" => m[1], "required" => true, "description" => m[2].strip }
        elsif (m = line.match(/\A\s+\[(\w[\w-]*)\]\s+(.*)/))
          { "name" => m[1], "required" => false, "description" => m[2].strip }
        end
      end
    end

    def extract_flags(text)
      section = extract_section(text, "FLAGS")
      return [] unless section

      flags = []
      # A flag line looks like:
      #   "  -d, --due string       Due date (YYYY-MM-DD)"
      #   "      --attach stringArray   Attach file (repeatable)"
      #   "  -h, --help             help for create"
      section.each_line do |line|
        next unless (m = line.match(/\A\s+(?:-(\w),\s+)?--([\w-]+)(?:\s+(\S+))?\s{2,}(.*)/))

        short, name, type_hint, desc = m[1], m[2], m[3], m[4].strip
        next if name == "help" # not useful as a tool parameter

        flags << {
          "name"        => name,
          "short"       => short,
          "type"        => map_type(type_hint),
          "description" => desc,
        }.compact
      end
      flags
    end

    def map_type(hint)
      case hint
      when nil               then "boolean"       # bare flag, no value
      when "stringArray"     then "array"
      when "int"             then "integer"
      when "bool"            then "boolean"
      else                        "string"
      end
    end

    # Extracts the body of a top-level section header like "FLAGS" or "ARGUMENTS"
    # up to the next top-level header or end of text.
    def extract_section(text, name)
      in_section = false
      out = +""
      text.each_line do |line|
        if line.start_with?(name) && line.strip == name
          in_section = true
          next
        end
        if in_section
          # Next section header: line starts at column 0 with uppercase word.
          break if line.match?(/\A[A-Z][A-Z ]+\z/) || line.match?(/\A[A-Z][A-Z ]+\n/)
          out << line
        end
      end
      in_section ? out : nil
    end
  end
end

File: src/basecamp_cli_mcp/help_parser.py
"""Parse `basecamp <group> <action> --help` output into a structured schema.

Used only by the generator — not on the runtime hot path.
"""

from __future__ import annotations

import re
from typing import TypedDict


class Positional(TypedDict):
    name: str
    required: bool
    description: str


class Flag(TypedDict, total=False):
    name: str
    short: str | None
    type: str
    description: str


class Parsed(TypedDict):
    summary: str
    positional: list[Positional]
    flags: list[Flag]


_REQUIRED_POS = re.compile(r"\A\s+<(\w[\w-]*)>\s+(.*)")
_OPTIONAL_POS = re.compile(r"\A\s+\[(\w[\w-]*)]\s+(.*)")
_FLAG_LINE = re.compile(r"\A\s+(?:-(\w),\s+)?--([\w-]+)(?:\s+(\S+))?\s{2,}(.*)")
_HEADER = re.compile(r"\A[A-Z][A-Z ]+\Z")


def parse(help_text: str) -> Parsed:
    return {
        "summary": _summary(help_text),
        "positional": _positional(help_text),
        "flags": _flags(help_text),
    }


def _summary(text: str) -> str:
    lines: list[str] = []
    for line in text.splitlines():
        if line.strip() == "USAGE":
            break
        if line.strip():
            lines.append(line.strip())
    return " ".join(lines)


def _positional(text: str) -> list[Positional]:
    section = _section(text, "ARGUMENTS")
    if section is None:
        return []
    out: list[Positional] = []
    for line in section.splitlines(keepends=True):
        m = _REQUIRED_POS.match(line)
        if m:
            out.append({"name": m.group(1), "required": True, "description": m.group(2).strip()})
            continue
        m = _OPTIONAL_POS.match(line)
        if m:
            out.append({"name": m.group(1), "required": False, "description": m.group(2).strip()})
    return out


def _flags(text: str) -> list[Flag]:
    section = _section(text, "FLAGS")
    if section is None:
        return []
    out: list[Flag] = []
    for line in section.splitlines(keepends=True):
        m = _FLAG_LINE.match(line)
        if not m:
            continue
        short, name, type_hint, desc = m.group(1), m.group(2), m.group(3), m.group(4).strip()
        if name == "help":
            continue
        flag: Flag = {"name": name, "type": _map_type(type_hint), "description": desc}
        if short:
            flag["short"] = short
        out.append(flag)
    return out


def _map_type(hint: str | None) -> str:
    if hint is None:
        return "boolean"
    if hint == "stringArray":
        return "array"
    if hint == "int":
        return "integer"
    if hint == "bool":
        return "boolean"
    return "string"


def _section(text: str, name: str) -> str | None:
    """Return body of a top-level section like FLAGS / ARGUMENTS, up to the next header."""
    in_section = False
    out: list[str] = []
    for line in text.splitlines(keepends=True):
        stripped = line.strip()
        if not in_section:
            if stripped == name:
                in_section = True
            continue
        # Next top-level header (all-caps line at col 0) ends the section.
        if _HEADER.match(stripped) and line[:1].isalpha() and line[:1].isupper() and line == line.lstrip():
            break
        out.append(line)
    return "".join(out) if in_section else None
